### PR TITLE
refactor(gateway): add type annotations

### PR DESCRIPTION
**Pull request by doublevcodes**
## Summary

This PR adds typing to all methods in `gateway.py` as raised in #816.

## Checklist

- [ ] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
  - [X] I have run `task pyright` and fixed the relevant issues.
- [X] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:72]: Shouldn't the return value be typed in the `__init__` methods?

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:137]: Why aren't these variables typed?

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:152]: Same here

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:306]: You're missing all of the attributes set in this method

**Comment by doublevcodes** [comment]
This is my first PR in a while, so my Python may be a little rusty!

Feel free to correct me and be critical in reviews, I've tried my best to type everything correctly.

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:349]: Missing type hint

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:481]: Use `dict[str, Any]` instead

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:490]: Untyped variables

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:545]: Untyped attributes

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:685]: Use the generic list instead of the normal one

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:687]: Untyped

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:807]: You missed `hook`

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:814]: Untyped

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:851]: You missed `hook`

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:889]: Untyped

**Review by EmreTech — COMMENTED** [review]
Otherwise, nice job!

**Comment by doublevcodes** [comment]
I think this should be good now!

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:686]: 1. Using `List` from typing is unnecessary since `from __future__ import annotations` is used in this file
2. You should be more specific with the type. As in, what type is each item in this list?

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:819]: Be more specific with this type hint. I.e. use the brackets after the type hint itself and specify parameters (if you're confused on what I mean, look at the [docs](https://docs.python.org/3/library/typing.html#callable) for more info)

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:859]: Be more specific with the `hook` parameter.

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:963]: Be more specific with the `data` parameter.

**Review by EmreTech — COMMENTED** [review]
Just a few more things

**Comment by doublevcodes** [comment]
All done!

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:400]: Be more specific with the type hints for the parameters `predicate` and `result`

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:812]: Be more specific with the type hint for parameter `hook`

**Review by EmreTech — APPROVED** [review]
Nvm, looks good

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:688]:
```suggestion
            activities: List[Activity] = []
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:708]:
```suggestion
        user_ids: Optional[List[int]] = None,
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:200]:
```suggestion
    def get_payload(self) -> Dict[str, Any]:
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:225]:
```suggestion
    def get_payload(self) -> Dict[str, int]:
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:482]:
```suggestion
        message: Dict[str, Any] = utils._from_json(msg)
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:490]:
```suggestion
        data: Dict[str, Any] = message["d"]
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:903]:
```suggestion
        data: Dict[str, Any] = msg["d"]
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:900]:
```suggestion
    async def received_message(self, msg: Dict[str, Any]) -> None:
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:922]:
```suggestion
    async def initial_connection(self, data: Dict[str, Any]) -> None:
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:967]:
```suggestion
    async def load_secret_key(self, data: Dict[str, Any]) -> None:
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:308]:
```suggestion
    def __init__(self, socket: aiohttp.ClientWebSocketResponse, *, loop: asyncio.AbstractEventLoop) -> None:
        self.socket: aiohttp.ClientWebSocketResponse= socket
```

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:37]:
```suggestion
from typing import TYPE_CHECKING, Awaitable, Callable, NoReturn, Optional, Union, Dict, List
```

**Review by teaishealthy — CHANGES_REQUESTED** [review]

**Comment by doublevcodes** [comment]
I've made the changes but what was the reason for using `typing.List` as opposed to `list`?
Thanks for the review!

**Comment by doublevcodes** [comment]
Sorry I didn't realise what a mess of the history this made, I was making the changes using GitHub mobile and thought they'd be made as one commit 😅

**Comment by EmreTech** [comment]
Don't worry, it'll be squashed into one commit anyways

File: nextcord/gateway.py
Comment (EmreTech) [inline nextcord/gateway.py:325]: Pyright Strict will give errors about accessing a private attribute for this type hint. It's fine to remove it.

**Review by EmreTech — COMMENTED** [review]
Just one more thing and this PR should be golden

**Review by EmreTech — APPROVED** [review]
Good job!

File: nextcord/gateway.py
Comment (Skelmis) [inline nextcord/gateway.py:137]: Based on it popping to `None`

```suggestion
        interval: Optional[float] = kwargs.pop("interval", None)
```

**Review by Skelmis — CHANGES_REQUESTED** [review]
Otherwise lgtm

File: nextcord/gateway.py
Comment (teaishealthy) [inline nextcord/gateway.py:142]:
```suggestion
        self.interval: Optional[float] = interval
```

**Review by teaishealthy — APPROVED** [review]

**Review by teaishealthy — APPROVED** [review]

File: nextcord/gateway.py
Comment (ooliver1) [inline nextcord/gateway.py:684]:
```suggestion
    ) -> None:
```

File: nextcord/gateway.py
Comment (ooliver1) [inline nextcord/gateway.py:814]: https://github.com/nextcord/nextcord/pull/830/files#diff-d090b4de22b790cbf918f41845d109812ebb342752113453a347203ada62a6f4R865
Could this `Callable` be specified fully

File: nextcord/gateway.py
Comment (ooliver1) [inline nextcord/gateway.py:825]:
```suggestion
    async def _default_hook(self, *args: Any) -> None:
```

File: nextcord/gateway.py
Comment (ooliver1) [inline nextcord/gateway.py:975]:
```suggestion
    async def poll_event(self) -> None:
```
yes it may raise but `NoReturn` tells type checkers that the code below the call will not run, which is unnecessary here since it may not happen

**Review by ooliver1 — CHANGES_REQUESTED** [review]

File: nextcord/gateway.py
Comment (doublevcodes) [reply nextcord/gateway.py:814]: I'm not sure what the callable would look like specified fully. Could I have some assistance?

Comment (ooliver1) [reply nextcord/gateway.py:814]: https://github.com/nextcord/nextcord/pull/830/files#diff-d090b4de22b790cbf918f41845d109812ebb342752113453a347203ada62a6f4R821

**Comment by doublevcodes** [comment]
I think GitHub's broken in terms of request reviews 😅

**Review by Skelmis — APPROVED** [review]

**Comment by doublevcodes** [comment]
Hey I think this can be merged now?

**Comment by toolifelesstocode** [comment]
> Hey I think this can be merged now?

This pull request requires 2 approving reviews from 2 people with write access(people with the green tick next to the approve). So the person who requested changes would have to approve it 🙂

**Review by DenverCoder1 — APPROVED** [review]

**Review by ooliver1 — APPROVED** [review]

**Pull request merged**
